pilot/pkg/model/endpointshards.go

```diff
@@ -0,0 +1,199 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"strings"
+	"sync"
+
+	"istio.io/istio/pilot/pkg/serviceregistry/provider"
+	"istio.io/istio/pilot/pkg/util/sets"
+	"istio.io/istio/pkg/cluster"
+)
+
+// shardRegistry is a simplified interface for registries that can produce a shard key
+type shardRegistry interface {
+	Cluster() cluster.ID
+	Provider() provider.ID
+}
+
+func NewShardKey(cluster cluster.ID, provider provider.ID) ShardKey {
+	return ShardKey(fmt.Sprintf("%s/%s", cluster, provider))
+}
+
+// ShardKeyFromRegistry computes the shard key based on provider type and cluster id.
+func ShardKeyFromRegistry(instance shardRegistry) ShardKey {
+	return NewShardKey(instance.Cluster(), instance.Provider())
+}
+
+// ShardKey is the key for EndpointShards made of a key with the format "cluster/provider"
+type ShardKey string
+
+func (sk ShardKey) Cluster() cluster.ID {
```

Comment:
This is called really often. Would it make sense to make ShardKey a `struct{cluster, provider}` so this is faster?